agent/php_user_instrument_wraprec_hash.h

```diff
@@ -0,0 +1,249 @@
+  const char *metavalue = nr_php_op_array_ ## zf_metaname(&zf->op_array); \
+  const size_t metavaluelen = nr_php_op_array_ ## zf_metaname ## _length(&zf->op_array); \
+  if (NULL != metavalue) { \
+    id->nr_metaname.is_set = true; \
+    id->nr_metaname.len = metavaluelen; \
+    set_meta_value(&id->nr_metaname, metavalue, method); \
+  } else { \
+    id->nr_metaname.is_set = false; \
+  } \
+} while (0)
+
+/* Cast zend_function to nruserfn_metadata - id will point to metadata in zf.
+ * Cast does not make any memory allocation but makes the code more readable */
+static inline void zf_as_id(nruserfn_metadata* id, const zend_function* zf) {
+  SET_META_COMMON(id, zf);
+  SET_META_STRING(id, filename, zf, file_name, NULL);
+  SET_META_STRING(id, scope, zf, scope_name, NULL);
+  SET_META_STRING(id, function_name, zf, function_name, NULL);
+}
+
+/* Copy zend_function metadata to nruserfn_metadata - id will have copy zf's
+ * metadata. Copy allocates memory that the caller must free. */
+static inline void zf_to_id(nruserfn_metadata* id, const zend_function* zf) {
+  SET_META_COMMON(id, zf);
+  SET_META_STRING(id, filename, zf, file_name, nr_strdup);
+  SET_META_STRING(id, scope, zf, scope_name, nr_strdup);
+  SET_META_STRING(id, function_name, zf, function_name, nr_strdup);
+}
+
+static inline void wraprec_metadata_set(nruserfn_metadata* id, const zend_function* zf) {
+ if (id->is_set) {
+    return;
+  }
+  zf_to_id(id, zf);
+  id->is_set = true;
+}
+
+static inline bool wraprec_streq(const zf_metadata_t *wr_id, const zf_metadata_t* zf_id) {
+
+  if (!wr_id->is_set || !zf_id->is_set) {
+    return false;
+  }
+
+  /* This is an assert - it's already been checked */
+  if (wr_id->len != zf_id->len) {
+    return false;
+  }
+
+  /* compare from the back */
+  for (int i = wr_id->len; i >= 0; i--) {
+    if (wr_id->value.ccp[i] != zf_id->value.ccp[i]) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+static inline bool wraprec_metadata_is_match(const nruserfn_t* wraprec, const zend_function* zf) {
+  const nruserfn_metadata *nr_id;
+  nruserfn_metadata zf_id = {};
+
+  if (nrunlikely(NULL == wraprec || NULL == zf)) {
+    return false;
+  }
+
+  nr_id = &wraprec->id;
+
+  /* cast zend_function to nruserfn_metadata - cast does not make 
+   * any memory allocation but makes the code more readable */
+  zf_as_id(&zf_id, zf);
+
+  if (nr_id->lineno != zf_id.lineno) {
+    return false;
+  }
+
+  if (nr_id->function_name.len != zf_id.function_name.len) {
+    return false;
+  }
+
+  if (nr_id->filename.len != zf_id.filename.len) {
+    return false;
+  }
+
+  if (nr_id->scope.len != zf_id.scope.len) {
+    return false;
+  }
+
+  if (!wraprec_streq(&nr_id->function_name, &zf_id.function_name)) {
+    return false;
+  }
+
+  if (!wraprec_streq(&nr_id->filename, &zf_id.filename)) {
+    return false;
+  }
+
+  /* No need to check scope - function has the same name is at the same location in the same file */
```

Comment:
On a second thought it may be better to leave scope matching in. See [this comment](https://github.com/newrelic/newrelic-php-agent/pull/587/files#r1034950492).